bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/BoschHttpClient.java

```diff
@@ -0,0 +1,139 @@
+
+public class BoschHttpClient extends HttpClient {
+
+    private static final int MAX_PAIR_TRIES = 12;
+    
+	private final Logger logger = LoggerFactory.getLogger(BoschHttpClient.class);
+
+	private String ipAddress;
+	private String systempassword;
+
+	public BoschHttpClient(String ipAddress, String systempassword, SslContextFactory sslContextFactory) {
+		super(sslContextFactory);
+		this.ipAddress = ipAddress;
+		this.systempassword = systempassword;
+	}
+
+	private String getCertFromSslContextFactory() throws KeyStoreException, CertificateEncodingException {
+		Certificate cert = this.getSslContextFactory().getKeyStore().getCertificate(BoschSslUtil.OSS_OPENHAB_BINDING);
+		return new String(Base64.getEncoder().encode(cert.getEncoded()));
+	}
+
+	public void checkAccessAndPairIfNecessary() throws InterruptedException {
+        // test if pairing is needed
+   		int counter = 0;
+       	boolean accessPossible = isAccessPossible();
+       	while (!accessPossible) {
+			// sleep some seconds after every try, except the first one
+   			if (counter > 0) {
+   				logger.info("Last Pairing failed, starting retry number {}/{} in few seconds", counter, MAX_PAIR_TRIES);
+       			Thread.sleep(15000);
+   			} else {
+   				logger.info("Pairing needed, because access to Bosch SmartHomeController not possible.");
+   			}
+   			// Timeout after max tries with an exception that pairing failed
+   			if (counter >= MAX_PAIR_TRIES)  {
+   				// all tries during the last MAX_PAIR_TRIES*1500 milliseconds failed
+   				throw new IllegalStateException("Pairing abourted, because it failed too many times!");
+   			}
+   			// try to pair if no access is possible 
+   			counter++;
+   			doPairing();
+   			// check access
+   			accessPossible = isAccessPossible();
+   		} 
+       
+	}
+
+	public Boolean isAccessPossible() {
+		try {
+			ContentResponse contentResponse = newRequest("https://" + ipAddress + ":8444/smarthome/devices").header("Content-Type", "application/json")
+					.header("Accept", "application/json").method(GET).send();
+			String content = contentResponse.getContentAsString();
+			logger.debug("Access check response complete: {} - return code: {}", content, contentResponse.getStatus());
+			return true;
+		} catch (InterruptedException | TimeoutException | ExecutionException e) {
+			logger.debug("Access check response failed!", e);
+			return false;
+		}
+	}
+
+
+	public boolean doPairing() {
+		logger.info("Starting pairing OpenHab Client with Bosch SmartHomeController!");
+		logger.info("Please press the Bosch SHC button until LED starts blinking");
+
+		ContentResponse contentResponse;
+		try {
+			String publicCert = getCertFromSslContextFactory();
+			logger.info("Pairing with Cert {}", publicCert);
+
+			// JSON Rest content
+			Map<String, String> items = new HashMap<>();
+			items.put("@type", "client");
+			// TODO Is a unique OpenHAB client ID needed or can we use the same name for all OpenHAB clients?
```

Comment:
I think I used the same one multiple times when I started developing the binding half a year ago, but I'm not 100% sure :-)

Reply:
I expect a naming problem if user runs multiple OpenHAB instacnes connected to the same SHC and I know that Bosch enforces a client starting with OSS in the [Terms and Conditions](https://github.com/BoschSmartHome/bosch-shc-api-docs#terms-and-conditions). 

- [ ] crosscheck with Bosch SHC terms of usage about client names

Reply:
From the Terms and Conditions:
> Using the API requires ... an individual Client ID

Therefore we need to find an ID for the current OpenHAB instance or generate a new UUID and store it.

Reply:
``ConfigConstants.getUserDataFolder()`` and ``InstanceUUID.get()`` combined together solve our issues: A unique cert inside a new JKS can be stored in "userdata/etc".
The path and the second password are not necessary anymore.

Reply:
If it's not a big thing I would recommend that we use a subfolder inside userdata/etc to put our JKS there. I just looked into the etc folder and it is already quite "crowded" in there ;)